tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <atomic>
+#include <csignal>
+#include <memory>
+#include <queue>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "tc/autotuner/genetic_search.h"
+#include "tc/autotuner/options_cache.h"
+#include "tc/autotuner/parameters.h"
+#include "tc/autotuner/utils.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/parser.h"
+
+namespace tc {
+namespace autotune {
+
+namespace detail {
+/**
+ * Internal harness to support multithreaded compilation and evaluation over
+ * various backends for a particular SearchStrategy.
+ * The SearchStrategy is a template parameter passed to the run method.
+ */
+template <typename Backend>
+class TuningHarness {
+ public:
+  using ExecutorType = typename Backend::ExecutorType;
+  using MappingOptionsType = typename Backend::MappingOptionsType;
+  using OptionsCacheType = OptionsCache<Backend>;
+
+  TuningHarness(
+      size_t maxPopulationSize,
+      lang::TreeRef tcTree,
+      const std::unordered_map<size_t, std::vector<const DLConstTensor*>>&
+          inputs,
+      std::unordered_map<size_t, std::vector<const DLTensor*>>& outputs,
+      const MappingOptionsType& baseMapping,
+      const TuningParameterFixer& fixedParams,
+      std::shared_ptr<OptionsCache<Backend>> optionsCache);
+
+  /// Runs a SearchStrategy
+  template <typename SearchStrategy>
+  void run(SearchStrategy& searchStrategy);
+
+  /// Once a signal has been caught, a flag is set and we terminate after the
+  /// current iteration is finished. This is apparently necessary for proper
+  /// python termination.
+  /// TODO: we should detect when we come from python and exit properly in C++.
+  void stopAfterCurrentIteration();
+
+ private:
+  /// Traverse one iteration of candidates in parallel and evaluate their
+  /// runtimes
+  template <typename SearchStrategy>
+  void runOneIteration(SearchStrategy& searchStrategy, size_t iteration);
+
+  /// Helper function to delegate compiling on the cpu to different threads
+  template <typename SearchStrategy>
+  void doCompile(SearchStrategy& searchStrategy);
+
+  /// Helper function to delegate running the compiled code to different
+  /// threads.
+  /// This function must be specialized per Backend
+  /// TODO: virtual inheritance?
```

Comment:
virtual inheritance makes me think of virtual base classes, which is not what was intended here